### PR TITLE
ignore bash built-in command exec

### LINKED ISSUE #1302: Xonsh breaks ssh-copy-id
I have Xonsh both on server and client as default shell, and this happens:

```
feanor@render_slave ~ $ ssh-copy-id feanor@10.42.0.1
/usr/bin/ssh-copy-id: INFO: Source of key(s) to be installed: "/home/feanor/.ssh/id_rsa.pub"
/usr/bin/ssh-copy-id: INFO: attempting to log in with the new key(s), to filter out any that are already installed
/usr/bin/ssh-copy-id: INFO: 1 key(s) remain to be installed -- if you are prompted now it is to install the new keys
feanor@10.42.0.1's password: 
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/xonsh/built_ins.py", line 472, in run_subproc
    **subproc_kwargs)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1541, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'exec'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/xonsh", line 9, in <module>
    load_entry_point('xonsh==0.3.4', 'console_scripts', 'xonsh')()
  File "/usr/local/lib/python3.5/dist-packages/xonsh/main.py", line 241, in main
    run_code_with_cache(args.command, shell.execer, mode='single')
  File "/usr/local/lib/python3.5/dist-packages/xonsh/codecache.py", line 182, in run_code_with_cache
    run_compiled_code(ccode, glb, loc, mode)
  File "/usr/local/lib/python3.5/dist-packages/xonsh/codecache.py", line 57, in run_compiled_code
    func(code, glb, loc)
  File "74c9f1de0f9d47f6f60e067c2c822ae0", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/xonsh/built_ins.py", line 607, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/usr/local/lib/python3.5/dist-packages/xonsh/built_ins.py", line 482, in run_subproc
    raise XonshError(e)
xonsh.tools.XonshError: xonsh: subprocess mode: command not found: exec
Did you mean one of the following?
    ex:     Command (/usr/bin/ex)
    xexec:  Alias
    xev:    Command (/usr/bin/xev)
    xgc:    Command (/usr/bin/xgc)
    jexec:  Command (/usr/bin/jexec)

No command 'exec' found, did you mean:
 Command 'pexec' from package 'pexec' (universe)
 Command 'gexec' from package 'gexec' (universe)
 Command 'kexec' from package 'kexec-tools' (main)
 Command 'jexec' from package 'openjdk-8-jre-headless' (main)
 Command 'jexec' from package 'default-jre' (main)
 Command 'jexec' from package 'openjdk-9-jre-headless' (universe)
 Command 'hexec' from package 'hexec' (universe)
feanor@render_slave ~ $ 

```

### PR DESCRIPTION
**Pull request by mitnk**
An attempt to fix #1302 , which I believe is caused by bash `exec` command. Though I'm not sure what's the correct way is.

### CODE REVIEW

File: xonsh/built_ins.py [hunk 1]
```diff
@@ -345,6 +345,12 @@ def run_subproc(cmds, captured=False):
 
     Lastly, the captured argument affects only the last real command.
     """
+    # exec is a bash built-in command and cannot be called by xonsh
+    if cmds and cmds[0] and cmds[0][0] == 'exec':
+        _cmds = list(cmds)
+        _cmds[0] = cmds[0][1:]
+        cmds = tuple(_cmds)
```
Comment (scopatz) [inline xonsh/built_ins.py]: I think that this is a bit of a hack.  It would be better to have an `exec` alias that called `run_subproc()`

**Comment by scopatz** [comment]
Thanks for putting this in @mitnk!

File: xonsh/built_ins.py (hunk 1 above)
Comment (mitnk) [inline xonsh/built_ins.py]: hmm, sounds a good idea, I will update it.

**Comment by mitnk** [comment]
@scopatz it's ready for another look.

**Comment by mitnk** [comment]
Unfortunately found that the current implementation (using `run_subproc()` series) will cause the piped commands stuck. Any suggestion to fix this?
```
$ exec echo foo | grep o
```

**Comment by mitnk** [comment]
And I just noticed that there's already an alias called [xexec](http://xon.sh/aliases.html#xexec), which is quite related the base exec. Any thoughts/concerns on this :)

**Comment by scopatz** [comment]
So is your problem solved by `aliases['exec'] = aliases['xexec']`?

**Comment by mitnk** [comment]
Ah, funny, I had tried xexec the piped command example and it replaces (disappears) this xonsh session (now I realize it's what expected), and I thought it's not what we need.

I just tried map `exec` to `xexec()` function in side `xonsh/aliases.py, it did fix my problem (though bare `aliases['exec'] = aliases['xexec']` won't work here).

However, one little issue differ from bash behavior:
1) open a fresh MacVim via `$ mvim <file-name>` inside xonsh session
2) quit xonsh completely.
3) type `Cmd + N` in MacVim window to open New window will not work, with report by apple I paste below.

Both the `xexec` and my current implement version has the same issue. but not when bash as the default shell.

```
Process:               Python [33811]
Path:                  /Library/Frameworks/Python.framework/Versions/3.5/Resources/Python.app/Contents/MacOS/Python
Identifier:            Python
Version:               3.5.0 (3.5.0)
Code Type:             X86-64 (Native)
Parent Process:        MacVim [33762]
Responsible:           Python [33811]
User ID:               501

Date/Time:             2016-06-29 09:09:25.626 +0800
OS Version:            Mac OS X 10.11.5 (15F34)
Report Version:        11
Anonymous UUID:        F4EAB527-CCBC-C246-3555-10D3F75B4D90

Sleep/Wake UUID:       9003B1BE-6AD2-4EE2-8025-D7540B327021

Time Awake Since Boot: 110000 seconds
Time Since Wake:       1600 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Application Specific Information:
abort() called

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff9acacf06 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff992d84ec pthread_kill + 90
2   libsystem_c.dylib             	0x00007fff89e666e7 abort + 129
3   org.python.python             	0x000000010011a831 Py_FatalError + 65
4   org.python.python             	0x000000010011c52e _Py_InitializeEx_Private + 1358
5   org.python.python             	0x0000000100138910 Py_Main + 1264
6   org.python.python             	0x0000000100000e32 0x100000000 + 3634
7   org.python.python             	0x0000000100000c84 0x100000000 + 3204

Thread 0 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000000  rbx: 0x0000000000000006  rcx: 0x00007fff5bfff448  rdx: 0x0000000000000000
  rdi: 0x0000000000000a0b  rsi: 0x0000000000000006  rbp: 0x00007fff5bfff470  rsp: 0x00007fff5bfff448
   r8: 0x00000000000000a9   r9: 0x0000000000000030  r10: 0x0000000008000000  r11: 0x0000000000000206
  r12: 0x00007fff77b32bd0  r13: 0x0000000100226c10  r14: 0x00007fff7a8e7000  r15: 0x00007fff5bfff4d8
  rip: 0x00007fff9acacf06  rfl: 0x0000000000000206  cr2: 0x00007fff77b30008
  
Logical CPU:     0
Error Code:      0x02000148
Trap Number:     133


Binary Images:
       0x100000000 -        0x100000ff7 +org.python.python (3.5.0 - 3.5.0) <B001A7B7-1E1A-D1B2-C16F-626D04F1860E> /Library/Frameworks/Python.framework/Versions/3.5/Resources/Python.app/Contents/MacOS/Python
       0x100003000 -        0x100212fff +org.python.python (3.5.0, [c] 2001-2015 Python Software Foundation. - 3.5.0) <CF159276-3178-8CCA-C918-5B81CD173154> /Library/Frameworks/Python.framework/Versions/3.5/Python
    0x7fff688b1000 -     0x7fff688e825f  dyld (360.22) <A468D85E-D8D6-3461-8C99-49D3B9ACFC63> /usr/lib/dyld
    0x7fff87983000 -     0x7fff879acfff  libc++abi.dylib (125) <DCCC8177-3D09-35BC-9784-2A04FEC4C71B> /usr/lib/libc++abi.dylib
    0x7fff87f00000 -     0x7fff87f00ff7  libunc.dylib (29) <DDB1E947-C775-33B8-B461-63E5EB698F0E> /usr/lib/system/libunc.dylib
    0x7fff88444000 -     0x7fff88449ff3  libunwind.dylib (35.3) <F6EB48E5-4D12-359A-AB54-C937FBBE9043> /usr/lib/system/libunwind.dylib
    0x7fff88786000 -     0x7fff88787fff  libsystem_blocks.dylib (65) <1244D9D5-F6AA-35BB-B307-86851C24B8E5> /usr/lib/system/libsystem_blocks.dylib
    0x7fff89e08000 -     0x7fff89e95fff  libsystem_c.dylib (1082.50.1) <B552D565-B798-3B9B-AE63-F623B42A5F01> /usr/lib/system/libsystem_c.dylib
    0x7fff8a003000 -     0x7fff8a056ff7  libc++.1.dylib (120.1) <8FC3D139-8055-3498-9AC5-6467CB7F4D14> /usr/lib/libc++.1.dylib
    0x7fff8a16e000 -     0x7fff8a176fff  libsystem_networkextension.dylib (385.40.36) <66095DC7-6539-38F2-95EE-458F15F6D014> /usr/lib/system/libsystem_networkextension.dylib
    0x7fff8d5fc000 -     0x7fff8d604fef  libsystem_platform.dylib (74.40.2) <29A905EF-6777-3C33-82B0-6C3A88C4BA15> /usr/lib/system/libsystem_platform.dylib
    0x7fff8d89f000 -     0x7fff8d8a7ffb  libsystem_dnssd.dylib (625.50.5) <4D10E12B-59B5-386F-82DA-326F18028F0A> /usr/lib/system/libsystem_dnssd.dylib
    0x7fff8ea8c000 -     0x7fff8eaa3ff7  libsystem_asl.dylib (323.50.1) <41F8E11F-1BD0-3F1D-BA3A-AA1577ED98A9> /usr/lib/system/libsystem_asl.dylib
    0x7fff8ef2e000 -     0x7fff8f299657  libobjc.A.dylib (680) <D55D5807-1FBE-32A5-9105-44D7AFE68C27> /usr/lib/libobjc.A.dylib
    0x7fff8f356000 -     0x7fff8f563fff  libicucore.A.dylib (551.51.3) <5BC80F94-C90D-3175-BD96-FF1DC222EC9C> /usr/lib/libicucore.A.dylib
    0x7fff8fcc7000 -     0x7fff8fccbfff  libcache.dylib (75) <9548AAE9-2AB7-3525-9ECE-A2A7C4688447> /usr/lib/system/libcache.dylib
    0x7fff90aa6000 -     0x7fff90ad5ffb  libsystem_m.dylib (3105) <08E1A4B2-6448-3DFE-A58C-ACC7335BE7E4> /usr/lib/system/libsystem_m.dylib
    0x7fff90b0f000 -     0x7fff90b38fff  libsystem_info.dylib (477.50.4) <FAA9226D-64DE-3769-A6D8-6CABA4B7FF4D> /usr/lib/system/libsystem_info.dylib
    0x7fff90d3d000 -     0x7fff90d42ff7  libmacho.dylib (875.1) <318264FA-58F1-39D8-8285-1F6254EE410E> /usr/lib/system/libmacho.dylib
    0x7fff91448000 -     0x7fff91475fff  libdispatch.dylib (501.40.12) <C7499857-61A5-3D7D-A5EA-65DCC8C3DF92> /usr/lib/system/libdispatch.dylib
    0x7fff91d74000 -     0x7fff91d75fff  libDiagnosticMessagesClient.dylib (100) <4243B6B4-21E9-355B-9C5A-95A216233B96> /usr/lib/libDiagnosticMessagesClient.dylib
    0x7fff9225a000 -     0x7fff922d1feb  libcorecrypto.dylib (335.50.1) <B5C05FD7-A540-345A-87BF-8E41848A3C17> /usr/lib/system/libcorecrypto.dylib
    0x7fff927c1000 -     0x7fff927d8ff7  libsystem_coretls.dylib (83.40.5) <C90DAE38-4082-381C-A185-2A6A8B677628> /usr/lib/system/libsystem_coretls.dylib
    0x7fff935ce000 -     0x7fff935d0fff  libsystem_coreservices.dylib (19.2) <1B3F5AFC-FFCD-3ECB-8B9A-5538366FB20D> /usr/lib/system/libsystem_coreservices.dylib
    0x7fff9461e000 -     0x7fff9463aff7  libsystem_malloc.dylib (67.40.1) <5748E8B2-F81C-34C6-8B13-456213127678> /usr/lib/system/libsystem_malloc.dylib
    0x7fff94f7d000 -     0x7fff94f84ff7  libcompiler_rt.dylib (62) <A13ECF69-F59F-38AE-8609-7B731450FBCD> /usr/lib/system/libcompiler_rt.dylib
    0x7fff95af4000 -     0x7fff95b5aff7  libsystem_network.dylib (583.50.1) <B52DAB73-92DC-3DA7-B9F4-B899D66445C1> /usr/lib/system/libsystem_network.dylib
    0x7fff95e30000 -     0x7fff95e31ffb  libremovefile.dylib (41) <552EF39E-14D7-363E-9059-4565AC2F894E> /usr/lib/system/libremovefile.dylib
    0x7fff96237000 -     0x7fff96239ff7  libsystem_configuration.dylib (802.40.13) <3DEB7DF9-6804-37E1-BC83-0166882FF0FF> /usr/lib/system/libsystem_configuration.dylib
    0x7fff9827e000 -     0x7fff9827eff7  libkeymgr.dylib (28) <8371CE54-5FDD-3CE9-B3DF-E98C761B6FE0> /usr/lib/system/libkeymgr.dylib
    0x7fff9827f000 -     0x7fff98290ff7  libz.1.dylib (61.20.1) <B3EBB42F-48E3-3287-9F0D-308E04D407AC> /usr/lib/libz.1.dylib
    0x7fff982fd000 -     0x7fff98773fff  com.apple.CoreFoundation (6.9 - 1258.1) <943A1383-DA6A-3DC0-ABCD-D9AEB3D0D34D> /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
    0x7fff992d2000 -     0x7fff992dbff7  libsystem_pthread.dylib (138.10.4) <3DD1EF4C-1D1B-3ABF-8CC6-B3B1CEEE9559> /usr/lib/system/libsystem_pthread.dylib
    0x7fff992f0000 -     0x7fff992f3ffb  libdyld.dylib (360.22) <CC088C2A-D407-33E7-A6B6-B06E0D4AD999> /usr/lib/system/libdyld.dylib
    0x7fff99cbc000 -     0x7fff99ccdff7  libsystem_trace.dylib (201.10.3) <F00E92E4-DBDA-3749-B5B3-0C3FBBABA1CB> /usr/lib/system/libsystem_trace.dylib
    0x7fff9ac88000 -     0x7fff9ac91ff3  libsystem_notify.dylib (150.40.1) <D48BDE34-0F7E-34CA-A0FF-C578E39987CC> /usr/lib/system/libsystem_notify.dylib
    0x7fff9ac96000 -     0x7fff9acb4ff7  libsystem_kernel.dylib (3248.50.21) <78E54D59-D2B0-3F54-9A4A-0A68D671F253> /usr/lib/system/libsystem_kernel.dylib
    0x7fff9ad23000 -     0x7fff9ad2eff7  libcommonCrypto.dylib (60075.50.1) <93732261-34B4-3914-B7A2-90A81A182DBA> /usr/lib/system/libcommonCrypto.dylib
    0x7fff9b239000 -     0x7fff9b23bff7  libquarantine.dylib (80) <0F4169F0-0C84-3A25-B3AE-E47B3586D908> /usr/lib/system/libquarantine.dylib
    0x7fff9b675000 -     0x7fff9b678fff  libsystem_sandbox.dylib (460.50.4) <150A9D3D-F69E-32F7-8C7B-8E72CAAFF7E4> /usr/lib/system/libsystem_sandbox.dylib
    0x7fff9b6b1000 -     0x7fff9b6b2ffb  libSystem.B.dylib (1226.10.1) <C5D09FE1-CC70-383E-AC27-18602F2EDEC4> /usr/lib/libSystem.B.dylib
    0x7fff9bcf9000 -     0x7fff9bd3fff7  libauto.dylib (186) <999E610F-41FC-32A3-ADCA-5EC049B65DFB> /usr/lib/libauto.dylib
    0x7fff9c2c0000 -     0x7fff9c2c8fff  libcopyfile.dylib (127) <A48637BC-F3F2-34F2-BB68-4C65FD012832> /usr/lib/system/libcopyfile.dylib
    0x7fff9c517000 -     0x7fff9c540ff7  libxpc.dylib (765.50.8) <54D1328E-054E-3DAA-89E2-375722F9D18F> /usr/lib/system/libxpc.dylib
    0x7fff9d3a2000 -     0x7fff9d3a2ff7  liblaunch.dylib (765.50.8) <834ED605-5114-3641-AA4D-ECF31B801C50> /usr/lib/system/liblaunch.dylib
    0x7fff9dfac000 -     0x7fff9dfadfff  libsystem_secinit.dylib (20) <32B1A8C6-DC84-3F4F-B8CE-9A52B47C3E6B> /usr/lib/system/libsystem_secinit.dylib

External Modification Summary:
  Calls made by other processes targeting this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by all processes on this machine:
    task_for_pid: 4307242
    thread_create: 0
    thread_set_state: 0

VM Region Summary:
ReadOnly portion of Libraries: Total=108.4M resident=0K(0%) swapped_out_or_unallocated=108.4M(100%)
Writable regions: Total=37.7M written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=37.7M(100%)
 
                                VIRTUAL   REGION 
REGION TYPE                        SIZE    COUNT (non-coalesced) 
===========                     =======  ======= 
Activity Tracing                  2048K        2 
Kernel Alloc Once                    4K        2 
MALLOC                            19.2M       12 
MALLOC guard page                   16K        4 
MALLOC_LARGE (reserved)            128K        2         reserved VM address space (unallocated)
VM_ALLOCATE                          4K        2 
__DATA                            3404K       47 
__LINKEDIT                        92.2M        5 
__TEXT                            16.2M       47 
__UNICODE                          552K        2 
__UNIXSTACK                       16.0M        2 
shared memory                        8K        3 
===========                     =======  ======= 
TOTAL                            149.6M      118 
TOTAL, minus reserved VM space   149.4M      118 
```

**Comment by mitnk** [comment]
Anyway I updated this PR to use `xexec` now.

**Comment by hugo-aa** [comment]
Maybe we could move forward with the PR to fix [issues users found](https://github.com/xonsh/xonsh/issues/1302), though [not completely](https://github.com/xonsh/xonsh/pull/1361#issuecomment-229231728).

**Comment by mitnk** [comment]
(Sorry, changing account ..)

Maybe we could move forward with this PR to fix [issues users found](https://github.com/xonsh/xonsh/issues/1302), though [not completely](https://github.com/xonsh/xonsh/pull/1361#issuecomment-229231728).

**Comment by laerus** [comment]
@mitnk when you open vim in a xonsh session and then quit xonsh, vim is supposed to quit aswell right? i believe there is no detach/nohup functionality in xonsh yet. maybe MacVim tries to access the already killed xonsh instance somehow and it complains that its not there anymore.

**Comment by mitnk** [comment]
@laerus when the shell quits, MacVim is not suppose to quit, at least IMO, or the cases in bash - `mvim` opens a new MacVim process separated from the shell.

**Comment by PaulReiber** [comment]
I'm not on a mac so can't test this myself.

Hugo, what I think you're saying is that mvim launches a separate window
with MacVim, which (with bash) stays live even when the parent shell exits
- and that's not happening with xonsh - the window's getting killed.   Is
that a correct description of the problem?

% pstree -pluga

please if you can, run the above (within a separate terminal window) to
compare/contrast the two scenarios - where mvim's running right and staying
running when the shell exits (shell=bash) and where it's exiting when it
shouldn't (shell=xonsh) (we don't need to see the whole tree - just the
tree branches associated w/ your current login(s)

Question is, whether mvim gets in a distinct process group, or if it's
still in xonsh's process group while it's running.

If the second one... that's possibly the problem since all of the processes
in the process group would all get intentionally terminated when the shell
exits.

If it's not the process group it could be mvim's still associated with the
xonsh's controlling terminal... but let's wait on what we can learn from
the above first, before diving into that.

Thanks!

Kind regards,


*Paul Reiber*http://about.me/reiber

**Comment by mitnk** [comment]
Hi @PaulReiber Thanks for your advices! Here is what I got.

Same result got when xonsh or bash as the default shell.

```
tab 1$ mvim .vimrc

tab 2$ ps ax | grep -i vim
37034   ??  Ss     0:00.17 /Applications/MacVim.app/Contents/MacOS/Vim -f -g .vimrc
37035   ??  S      0:00.40 /Applications/MacVim.app/Contents/MacOS/MacVim -MMNoWindow yes

tab 3$ pstree -p 37034
-+= 00001 root /sbin/launchd
 \-+= 37034 mitnk /Applications/MacVim.app/Contents/MacOS/Vim -f -g .vimrc
   \--= 37035 mitnk /Applications/MacVim.app/Contents/MacOS/MacVim -MMNoWindow yes
```

 The vim process is not in the tree that contains xonsh/bash.

bash:
```
$ pstree -s bash
-+= 00001 root /sbin/launchd
 |-+= 00215 root /usr/sbin/cron
 | \-+- 00490 root /usr/sbin/cron
 |   \--= 00491 mitnk (bash)
 \-+= 02446 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2
   |-+= 37364 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fp mitnk
   | \-+= 37365 root login -fp mitnk
   |   \--= 37366 mitnk -bash
   |-+= 37393 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fp mitnk
   | \-+= 37394 root login -fp mitnk
   |   \--= 37395 mitnk -bash
   |-+= 37419 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fp mitnk
   | \-+= 37420 root login -fp mitnk
   |   \--= 37421 mitnk -bash
   \-+= 37473 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fp mitnk
     \-+= 37474 root login -fp mitnk
       \-+= 37475 mitnk -bash
         \-+= 37515 mitnk pstree -s bash
           \--- 37516 root ps -axwwo user,pid,ppid,pgid,command

$ pstree -s vim
-+= 00001 root /sbin/launchd
 \-+= 37389 mitnk /Applications/MacVim.app/Contents/MacOS/Vim -f -g .vimrc
   \--= 37390 mitnk /Applications/MacVim.app/Contents/MacOS/MacVim -MMNoWindow yes
```

xonsh:
```
$ pstree -s xonsh
-+= 00001 root /sbin/launchd
 \-+= 37556 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2
   |-+= 37559 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fp mitnk
   | \-+= 37560 root login -fp mitnk
   |   \--- 37561 mitnk /Library/Frameworks/Python.framework/Versions/3.5/Resources/Python.app/Contents/MacOS
   \-+= 37574 mitnk /Applications/iTerm.app/Contents/MacOS/iTerm2 --server login -fp mitnk
     \-+= 37575 root login -fp mitnk
       \-+- 37576 mitnk /Library/Frameworks/Python.framework/Versions/3.5/Resources/Python.app/Contents/MacOS
         \-+= 37581 mitnk /usr/local/bin/pstree -s xonsh
           \--- 37582 root ps -axwwo user,pid,ppid,pgid,command

$ pstree -s vim
-+= 00001 root /sbin/launchd
 \-+= 37572 mitnk /Applications/MacVim.app/Contents/MacOS/Vim -f -g .vimrc
   \--= 37573 mitnk /Applications/MacVim.app/Contents/MacOS/MacVim -MMNoWindow yes
```

**Comment by astronouth7303** [comment]
Please resolve this! I found that slim's default configuration also uses `exec`.

**Comment by mitnk** [comment]
@astronouth7303 does this PR fix your problem?

**Comment by scopatz** [comment]
I think that this would be a good candidate for an officially supported sh-compatibility xontrib. I am skeptical of putting this kind of change in as a default, but that would be mitigated as a xontrib - at least initially.

**Comment by hugo-aa** [comment]
@scopatz ok, I'll try to add such a xontrib.

**Comment by mitnk** [comment]
I can try add such a xontrib.

**Comment by scopatz** [comment]
Yeah, I think that all you need to do is to have a single line file in the xontrib dir called `sh.py` for now.

**Comment by astronouth7303** [comment]
You missed the opportunity for another x pun: `xompat`.

**Comment by scopatz** [comment]
I thought about it, but I think it is more important to be clear.

**Comment by hugo-aa** [comment]
@scopatz 

> I think that all you need to do is to have a small file in the xontrib dir called sh.py for now.

I found issue in this solution - When I open new MacVim window, it calls `['exec', '/Applications/MacVim.app/Contents/MacOS/Vim']` at an external xonsh env where this new xontrib didn't get load. Any ideas on this?

**Comment by astronouth7303** [comment]
How do you load the xontrib?

**Comment by hugo-aa** [comment]
@astronouth7303 I did load the new added xontrib in my .xonshrc - so exec works on my shell sessions. But when MacVim calls exec outside of the shell sessions where xonshrc didn't get load.

**Comment by astronouth7303** [comment]
Did you try setting it in you xonfig?

**Comment by hugo-aa** [comment]
@astronouth7303 you mean `~/.config/xonsh/config.json`? just had a try, didn't work either

**Comment by scopatz** [comment]
@mitnk - can you update this PR so that we can see what you are trying to run, please? Thanks!

**Comment by mitnk** [comment]
updated, thanks.

**Comment by astronouth7303** [comment]
Yeah, loading this xontrib in my `.xonshrc` doesn't allow `xonsh -c "exec ..."` invocations, which is probably the biggest impetus to this change.

**Comment by mitnk** [comment]
> the underlying problem is that both sh and Python define exec and it isn't entirely clear how safe it is for both to be defined. [[source](https://github.com/xonsh/xonsh/issues/1133#issuecomment-234872574)]

Yeah, this is the trade off we need to make. IMO, we should make sh stuff prioritised - xonsh is first a shell that need to do more things than run Python code. And in this particular case, `exec` is rare used in Python daily programming.

**Comment by mitnk** [comment]
My proposal is making `exec` a built-in pre-loaded xonsh alias (as xexec) that overwrite the `exec` statement in Python. Sounds not a good idea, but may be the right thing to do ;)

**Comment by scopatz** [comment]
OK, I am just going to dump all of my exec thoughts, which is probably long overdue.  

The first is that the mantra that “Python always wins” has served xonsh well as a guiding light. I am personally strongly against abandoning this for many reasons that I’ll spare this discussion. So let’s assume that Python always wins.  The problem with exec is that is that it mixes forms.  Let’s define those forms so as to remove ambiguity:

**exec-as-a-function**

```
exec(code[, globals[, locals]])
```

**exec-as-a-statement**

```
exec code[, globals[, locals]]
```

**exec-as-a-command**

```
![exec code]
```

**use case**

The first thing to note is xonsh, obviously, makes try to make it so that the command form can be expressed as statement form. In fact, the only use case for exec as a command is also for it to be statement.  That is, other programs expect sh-like syntax where `exec command` is possible, without jumping through hoops like `![exec command]` or `xexec command`.

**Python 2 vs 3**

In Python 2, exec was purposefully ambiguous between function form and statement form. Because Python always wins, that would have stopped this here.  However, in Python 3 they removed the statement form and only provide the builtin exec() function. This means that xonsh in Python 3 can provide an exec alias that is useable as a statement, if we want.  

The concern is that if we add this by default, this is effectively killing Python 2 support.  Now maybe this is already a dead idea, but it seemed more plausible when these issues first started cropping up. Or maybe in a Python 2 compatible version of xonsh, they would have to drop this alias.  Thoughts?

**What should exec do?**

Now that we know that we can have an exec statement/command, what work should it perform. I see three reasonable possibilities for what exec can mean:

* execute xonsh code (what `execx does now`)
* execute pure Python code (what `exec` does now, because Python always wins)
* execute a command, replacing the current xonsh instance (implemented with `os.execvpe()`, similar to what sh-based langs do, what `xexec` does now).

Exec-as-a-function currently only runs pure Python code.  However, I think it would be great if we could figure out a way to replace this with execx() eventually. No one has figured out the boot strapping, yet. If you to do `builtins.exec = execx` now things break amazingly. 

Exec-as-a-statement / exec-as-a-command from an internal consistency standpoint could also run xonsh code. This would be especially compelling if exec-as-a-function ran xonsh code.  However, the proposal is that exec-as-a-command will only launch other commands, and replace the process. This is implemented in xexec and is what sh-langs implement for their execs.

So all of the above being said, there are two competing visions for the endgame for this.  In the first:

* exec in python mode runs pure python code
* exec in subprocess mode runs pure commands
* execx in python and/or subprocess mode runs xonsh code

In the second vision:

* exec in python mode runs xonsh code
* exec in subprocess mode runs xonsh code

Now perhaps the second vision is redundant syntax for the same operation and should be avoided on those grounds.  However, I don’t feel like the first vision is something we can sneak in with a single alias.  It needs to be very well documented and delineated.  I think this is especially true since the number of people coming from Python 2 who may expect that exec-as-a-statement executes Python code is very high. 

So in summary, I think the trade-off for vision 1 is user confusion and potential Python 2 support for better interoperability with external applications that expect sh-lang syntax. 

As stated, we might not care about Python 2. And confusion, when it arises can be mitigated with documentation. 

I am very interested everyone’s thoughts on this.  I can see it causing problems no matter what we do or don’t do.

**Comment by astronouth7303** [comment]
Because of the prevalence of `exec` in shells, and the clear expectation of tools for its presence, I think this might be a special case worth examining? (Even against the zen of python).

I prefer scenario one myself. If there's some way for a word to be overloaded as both a command and a Python object, that would be cool.

As far as Python 2 support, I would say drop it. As a stand-alone application and not a library, there are much fewer cases that it's necessary. And seeing as you're building a xontrib community, Python 2 support will either fragment that community, or place an undue burden on xontributors. Not to mention complicate your own maintenance.

**Comment by scopatz** [comment]
> I prefer scenario one myself. If there's some way for a word to be overloaded as both a command and a Python object, that would be cool.

There is. This is a core part of xonsh already :)

> As far as Python 2 support, I would say drop it.

Yeah... we are signatories to the Python 3 statement https://python3statement.github.io/

**Comment by astronouth7303** [comment]
If `exec

**Comment by scopatz** [comment]
I feel like we are having this discussion because it is not clear what exec should do in these situations.  Even accepting the first vision, none of the fixes that have come through make it clear to users what exec does.  At a minimum there needs to be a section in the docs somewhere so that we can point users to it.  I do think that this issue has a certain amount of inherent confusion. This discussion is meant to resolve that.

**Comment by scopatz** [comment]
It is not about the implementation, which is easy. It is about the usability and managing expectations.  I hope that makes sense :).

**Comment by astronouth7303** [comment]
Yes. The current state of affairs is the most pure. But I believe its been amply demonstrated that some form of support for shell `exec` is necessary in the real world. (There's been three specific examples of external projects expecting the user's shell to handle `exec`, and my bet is that there's more hiding.)

The opt-in xontrib approach might be the best way to manage expectations. Even if 90% of people have it enabled, they still had to enable it. However, if xonsh gains in popularity, there will probably be some amount of people copy/pasting example `.xonshrc`s without really understanding what they're doing (as easy as a `.xonshrc` is).

On the flip side, having xonsh break things by default is not good. My experience with slim shows that this problem will crop up in unexpected places and may not be obvious what broke, and that a shell needs to be reliable and not break things. (Granted, I'm pretty far off the beaten path, but I'm sure it'll crop up somewhere, and it may not be obvious that the shell is to blame.)

**Comment by scopatz** [comment]
Yeah, I agree.  Having written the above, I am probably most in favor of vision 1 coming stock in xonsh (ie not in a xontrib). However, I don't want other people to be swayed by that too deeply.  And to reiterate, I think vision 1 takes more than just a single line change to aliases.py. I think it needs documentation, and if someone could come up with a test for it too that would be great!  I'd welcome a PR that did that.

**Comment by gforsyth** [comment]
I am unmoved by Python 2 and don't think anyone is going to add a py27 parser.  I agree that the `exec` of vision one is a chimera and will need to be well documented.  And there probably will be confusion.  That said, we state everywhere that xonsh is a Python3.4+-ish superset and so users should not be too taken aback when we don't support Python2.7 (maybe?).    

I don't have strong feelings on whether or not it's a xontrib or stock, so I'll defer to @scopatz on that.

**Comment by scopatz** [comment]
> I don't have strong feelings on whether or not it's a xontrib or stock, so I'll defer to @scopatz on that.
 Show all checks

Well I think it is failing as a xontrib right now in some cases because of we don't load rc or config files with our headless shell.  So stock kinda wins on functionality :)

**Comment by mitnk** [comment]
Thanks for the detailed information here! Still consuming them, but, yeah, let's drop the idea of supporting Python 2 - Supporting Py 2 is do evil to the community :)

**Comment by scopatz** [comment]
Sot it seems like those of us who have commented have converged on vision 1.  So what this needs now is an implementation. That can be done either in this PR or in another one. Whoever wants to tackle this, feel free!

**Comment by mitnk** [comment]
Thank you so much everyone here. I'm closing this PR, and will open a new one soon to try to implement vision 1 - if no one more speedy :)

**Comment by scopatz** [comment]
Thanks @mitnk!
